File: haskell/HOASCBV.hs
{-# language BlockArguments, BangPatterns #-}

module HOASCBV where

import Data.Time.Clock
import Control.Monad
import System.IO

data Val = VLam !(Val -> Val) | VVar !Int | VApp !Val !Val
data Tm  = Lam !Tm | App !Tm !Tm | Var !Int deriving Show

infixl 8 $$
($$) :: Val -> Val -> Val
VLam f $$ a = seq a (f a)
v      $$ a = VApp v a
{-# inline ($$) #-}

n2  = VLam \s -> VLam \z -> s $$ (s $$ z)
n5  = VLam \s -> VLam \z -> s $$ (s $$ (s $$ (s $$ (s $$ z))))
mul = VLam \a -> VLam \b -> VLam \s -> VLam \z -> a $$ (b $$ s) $$ z
suc n = VLam \s -> VLam \z -> s $$ (n $$ s $$ z)

n10      = mul $$ n2    $$ n5
n10b     = mul $$ n5    $$ n2
n20      = mul $$ n2    $$ n10
n20b     = mul $$ n2    $$ n10b
n21      = suc n20
n21b     = suc n20b
n22      = suc n21
n22b     = suc n21b
n100     = mul $$ n10   $$ n10
n100b    = mul $$ n10b  $$ n10b
n10k     = mul $$ n100  $$ n100
n10kb    = mul $$ n100b $$ n100b
n100k    = mul $$ n10k  $$ n10
n100kb   = mul $$ n10kb $$ n10b
n1M      = mul $$ n10k  $$ n100
n1Mb     = mul $$ n10kb $$ n100b
n5M      = mul $$ n1M   $$ n5
n5Mb     = mul $$ n1Mb  $$ n5
n10M     = mul $$ n1M   $$ n10
n10Mb    = mul $$ n1Mb  $$ n10b
n20M     = mul $$ n10M  $$ n2
n20Mb    = mul $$ n10Mb $$ n2
n100M    = mul $$ n10M  $$ n10
n100Mb   = mul $$ n10Mb $$ n10b

leaf     = VLam \l -> VLam \n -> l
node     = VLam \t1 -> VLam \t2 -> VLam \l -> VLam \n -> n $$ t1 $$ t2
fullTree = VLam \n -> n $$ (VLam \t -> node $$ t $$ t) $$ leaf

quote :: Int -> Val -> Tm
quote !d (VVar x)   = Var (d - x - 1)
quote  d (VApp t u) = App (quote d t) (quote d u)
quote  d (VLam t)   = Lam (quote (d + 1) (t (VVar d)))

quote0 = quote 0

conv :: Int -> Val -> Val -> Bool
conv !d t u = case (t, u) of
  (VVar x,   VVar x'   ) -> x == x'
  (VApp t u, VApp t' u') -> conv d t t' && conv d u u'
  (VLam t,   VLam t'   ) -> let v = VVar d in conv (d + 1) (t v) (t' v)
  _                      -> False

conv0 = conv 0

force :: Tm -> Bool
force Lam{} = True
force _     = False

natToInt :: Tm -> Int
natToInt (Lam (Lam t)) = go 0 t where
  go acc (App _ t) = go (acc + 1) t
  go acc _         = acc
natToInt _ = 0

timed :: Show b => String -> Int -> (a -> IO b) -> a -> IO ()
timed msg times act a = do
  putStrLn msg
  putStrLn $ "Iterations: " ++ show times
  t0 <- getCurrentTime
  replicateM_ times $ do
    b <- act a
    putStr $ show b ++ " "
  t1 <- getCurrentTime
  putStrLn $ "\nAverage time: " ++ show (diffUTCTime t1 t0 / fromIntegral times)
{-# noinline timed #-}


bench :: IO ()
bench = do
  hSetBuffering stdout NoBuffering
  timed "Nat 5M conversion"     20 (pure . conv0 n5M) n5Mb
  timed "Nat 5M normalization"  20 (pure . force . quote0) n5M
  timed "Nat 10M conversion"    20 (pure . conv0 n10M) n10Mb
  timed "Nat 10M normalization" 20 (pure . force . quote0) n10M
  timed "Tree 2M conversion"    20 (pure . conv0 (fullTree $$ n20)) (fullTree $$ n20b)
  timed "Tree 2M normalization" 20 (pure . force . quote0) (fullTree $$ n20)
  timed "Tree 4M conversion"    20 (pure . conv0 (fullTree $$ n21)) (fullTree $$ n21b)
  timed "Tree 4M normalization" 20 (pure . force . quote0) (fullTree $$ n21)
  timed "Tree 8M conversion"    20 (pure . conv0 (fullTree $$ n22)) (fullTree $$ n22b)
  timed "Tree 8M normalization" 20 (pure . force . quote0) (fullTree $$ n22)

File: ocaml/HOASCBV.ml

open Format

type value = VVar of int | VApp of value * value | VLam of (value -> value)
type tm = Var of int | App of tm * tm | Lam of tm

let [@inline] ap t u = match t with VLam f -> f u | t -> VApp(t, u)
let [@inline] ap2 t u v = ap (ap t u) v

let n2  = VLam (fun s -> VLam (fun z -> ap s (ap s z)))
let n5  = VLam (fun s -> VLam (fun z -> ap s (ap s (ap s (ap s (ap s z))))))
let mul = VLam (fun a -> VLam (fun b -> VLam (fun s -> VLam (fun z -> ap (ap a (ap b s)) z))))
let suc n  = VLam (fun s -> VLam (fun z -> ap s (ap2 n s z)))
let n10    = ap2 mul n2 n5
let n10b   = ap2 mul n5 n2
let n20    = ap2 mul n2 n10
let n20b   = ap2 mul n2 n10b
let n21    = suc n20
let n21b   = suc n20b
let n22    = suc n21
let n22b   = suc n21b
let n100   = ap2 mul n10  n10
let n1000  = ap2 mul n100 n10
let n100b  = ap2 mul n10b n10b
let n10k   = ap2 mul n100 n100
let n10kb  = ap2 mul n100b n100b
let n100k  = ap2 mul n10k n10
let n100kb = ap2 mul n10k n10
let n1M    = ap2 mul n10k n100
let n1Mb   = ap2 mul n10kb n100b
let n2M    = ap2 mul n1M  n2
let n2Mb   = ap2 mul n1Mb n2
let n5M    = ap2 mul n1M  n5
let n5Mb   = ap2 mul n1Mb  n5
let n10M   = ap2 mul n1M  n10
let n10Mb  = ap2 mul n1Mb n10b

let leaf = VLam(fun l -> VLam(fun n -> l))
let node = VLam(fun t1 -> VLam(fun t2 -> VLam(fun l -> VLam(fun n ->
                ap2 n t1 t2))))
let fullTree = VLam(fun n -> ap2 n (VLam(fun t -> ap2 node t t)) leaf)

let rec quote l v =
  match v with
    | VVar x      -> Var (l - x - 1)
    | VLam t      -> Lam (quote (l + 1) (t (VVar(l))))
    | VApp (t, u) -> App (quote l t, quote l u)

let quote0 = quote 0

let rec conv (d:int) l r =
    match (l, r) with
    | (VVar(x), VVar(y)) -> x = y
    | (VApp(l1, l2), VApp(r1, r2)) -> conv d l1 r1 && conv d l2 r2
    | (VLam(l), VLam(r)) -> conv (d + 1) (l(VVar(d))) (r(VVar(d)))
    | _ -> false

let conv0 = conv 0

let natToInt (t : tm) : int =
  match t with
    | Lam (Lam t) ->
        let acc : int ref = ref 0 in
        let u   : tm ref  = ref t in
        let upd() = match !u with App(_, u') -> u := u'; true
                                | _          -> false in
        while upd() do
          acc := !acc + 1
        done;
        !acc
    | _ -> 1

let force (t : tm) =
  match t with Lam(_) -> true
             | _      -> false

let [@inline never] timed (msg:string) (times: int) (act: unit -> bool) =
  printf "%s\n" msg;
  printf "Iterations: %d\n" times;
  let t1 = Sys.time() in
  for i = 1 to times do
    printf "%B\n" (act())
  done;
  let t2 = Sys.time() in
  printf "\nAverage time: %fs\n" ((t2 -. t1) /. float_of_int times)

let () =
  (* printf "%d\n" (natToInt(quote0(n10M))); *)

  timed "Nat 5M conversion"      5 (fun _ -> conv0 n5M n5Mb);
  timed "Nat 5M normalization"   5 (fun _ -> force (quote0 n5M));
  timed "Nat 10M conversion"     5 (fun _ -> conv0 n10M n10Mb);
  timed "Nat 10M normalization"  5 (fun _ -> force (quote0 n10M));
  timed "Tree 2M conversion"     5 (fun _ -> conv0 (ap fullTree n20) (ap fullTree n20b));
  timed "Tree 2M normalization"  5 (fun _ -> force (quote0 (ap fullTree n20)));
  timed "Tree 4M conversion"     5 (fun _ -> conv0 (ap fullTree n21) (ap fullTree n21b));
  timed "Tree 4M normalization"  5 (fun _ -> force (quote0 (ap fullTree n21)));
  timed "Tree 8M conversion"     5 (fun _ -> conv0 (ap fullTree n22) (ap fullTree n22b));
  timed "Tree 8M normalization"  5 (fun _ -> force (quote0 (ap fullTree n22)))
